rollup/l1_state_tracker/l1_tracker.go

```diff
@@ -0,0 +1,222 @@
+package l1_state_tracker
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/scroll-tech/go-ethereum/core/types"
+	"github.com/scroll-tech/go-ethereum/event"
+	"github.com/scroll-tech/go-ethereum/log"
+	"github.com/scroll-tech/go-ethereum/rollup/sync_service"
+	"github.com/scroll-tech/go-ethereum/rpc"
+)
+
+type L1Tracker struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	client           sync_service.EthClient
+	lastSyncedHeader *types.Header
+	subList          headerSubList
+	scope            event.SubscriptionScope
+}
+
+const (
+	// defaultSyncInterval is the frequency at which we query for new chain head
+	defaultSyncInterval = 12 * time.Second
+)
+
+func NewL1Tracker(ctx context.Context, l1Client sync_service.EthClient) (*L1Tracker, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	l1Tracker := &L1Tracker{
+		ctx:    ctx,
+		cancel: cancel,
+
+		client: l1Client,
+	}
+	l1Tracker.Start()
+	return l1Tracker, nil
+}
+
+func (t *L1Tracker) headerByDepth(depth ChainDepth, latestNumber *big.Int) (*types.Header, error) {
+	var blockNumber *big.Int
+	switch depth {
+	case LatestBlock:
+		blockNumber = big.NewInt(int64(rpc.LatestBlockNumber))
+	case SafeBlock:
+		blockNumber = big.NewInt(int64(rpc.SafeBlockNumber))
+	case FinalizedBlock:
+		blockNumber = big.NewInt(int64(rpc.FinalizedBlockNumber))
+	default:
+		blockNumber = big.NewInt(0).Sub(latestNumber, big.NewInt(int64(depth)))
+	}
+	header, err := t.client.HeaderByNumber(t.ctx, blockNumber)
+	if err != nil {
+		return nil, err
+	}
+	return header, nil
+}
+
+func (t *L1Tracker) newHead(header *types.Header) {
+	t.lastSyncedHeader = header
+	t.subList.stopSending()
+	t.subList.sendNewHeads(t.headerByDepth, header)
+}
+
+func (t *L1Tracker) syncLatestHead() error {
+	header, err := t.client.HeaderByNumber(t.ctx, big.NewInt(int64(rpc.LatestBlockNumber)))
+	if err != nil {
+		return err
+	}
+	if !header.Number.IsInt64() {
+		return fmt.Errorf("received unexpected block number in L1Client: %v", header.Number)
+	}
+	// sync is continuous
+	if t.lastSyncedHeader != nil || header.ParentHash != t.lastSyncedHeader.Hash() {
+		t.newHead(header)
+	} else { // reorg happened or some blocks were not synced
+		// todo: clear cache
+		t.newHead(header)
+	}
+	return nil
+}
+
+func (t *L1Tracker) Subscribe(channel HeaderChan, depth ChainDepth) event.Subscription {
```

Comment:
Let's try to simplify by with putting more burden on the subscriber. We should require the subscriber to **not block** and then we can do something like this:
`Subscribe(depth ChainDepth, callback func(lastHead, newHead uint64, reorg bool)) unsubscribe func()`. This will also make testing much easier as we can sequentially test stuff